crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1054,18 +1073,24 @@ impl<'db> TypeInferenceBuilder<'db> {
         builtins_symbol_ty_by_name(self.db, "dict").instance()
     }
 
+    /// Infer the type of the `iter` expression of the first comprehension.
+    fn infer_first_comprehension_iter(&mut self, comprehensions: &[ast::Comprehension]) {
+        let mut generators_iter = comprehensions.iter();
+        let Some(first_generator) = generators_iter.next() else {
+            unreachable!("Expression must contain at least one generator");
+        };
+        self.infer_expression(&first_generator.iter);
+    }
+
     fn infer_generator_expression(&mut self, generator: &ast::ExprGenerator) -> Type<'db> {
         let ast::ExprGenerator {
             range: _,
-            elt,
+            elt: _,
             generators,
             parenthesized: _,
         } = generator;
 
-        self.infer_expression(elt);
-        for generator in generators {
-            self.infer_comprehension(generator);
-        }
+        self.infer_first_comprehension_iter(generators);
 
         // TODO generator type
         Type::Unknown
```

Comment:
We could update this to return `typing.Generator`, and similarly below for `builtins.dict`, `builtins.list`, `builtins.set`. It should be a fairly simple change to resolve the type from typeshed.

But I don't feel strongly about whether we do that in this PR or not, since we'd still have a TODO to make it generic and properly infer the contained type, once we have generic types.

Reply:
> We could update this to return `typing.Generator`, and similarly below for `builtins.dict`, `builtins.list`, `builtins.set`. It should be a fairly simple change to resolve the type from typeshed.

Generators are actually somewhat interesting because do you infer the abstract type (`typing.Generator`) or the concrete type (`types.GeneratorType`)? I think mypy does the former and it causes issues... can't remember what pyright does...

Reply:
Oh, yeah... I think probably the concrete type (`types.GeneratorType`) is better, since in this case we know for sure it is the built-in Generator type, not something else that just implements the interface.